src/main/java/ch/njol/skript/expressions/ExprFormatNumber.java

```diff
@@ -0,0 +1,132 @@
+import ch.njol.skript.lang.Literal;
+import ch.njol.skript.lang.SkriptParser.ParseResult;
+import ch.njol.skript.lang.VariableString;
+import ch.njol.skript.util.Getter;
+import ch.njol.util.Kleenean;
+import org.bukkit.event.Event;
+import org.eclipse.jdt.annotation.Nullable;
+
+import java.text.DecimalFormat;
+
+@Name("Formatted Number")
+@Description(
+	"Converts numbers to human-readable format. By default, '###,###' (e.g. '123,456,789') will be used. For reference, see this "
+		+ "<a href=\"https://docs.oracle.com/javase/7/docs/api/java/text/DecimalFormat.html\" target=\"_blank\">article</a>."
+)
+@Examples({
+	"command /formatnumber <number>:",
+		"\taliases: fn",
+		"\ttrigger:",
+			"\t\tsend \"Formatted: %formatted arg-1%\" to sender"
+})
+@Since("INSERT VERSION")
+public class ExprFormatNumber extends PropertyExpression<Number, String> {
+	
+	private static final DecimalFormat DEFAULT_FORMAT = new DecimalFormat("###,###");
+	
+	static {
+		Skript.registerExpression(ExprFormatNumber.class, String.class, ExpressionType.PROPERTY,
+				"%numbers% formatted [human-readable] [(with|as) %-string%]",
+				"[human-readable] formatted %numbers% [(with|as) %-string%]");
+	}
+
+	@SuppressWarnings("NotNullFieldNotInitialized")
+	private DecimalFormat format;
+
+	@Nullable
+	private Expression<? extends String> customFormat;
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		setExpr((Expression<? extends Number>) exprs[0]);
+		customFormat = (Expression<? extends String>) exprs[1];
+
+		boolean isSimpleString = customFormat instanceof VariableString && ((VariableString) customFormat).isSimple();
+		// if literal or is simple variable string then we can get the value at init and check
+		if (customFormat instanceof Literal || isSimpleString) {
+			String customFormatValue;
+			if (isSimpleString) {
+				customFormatValue = ((VariableString) customFormat).toString(null);
+			} else {
+				customFormatValue = ((Literal<String>) customFormat).getSingle();
+			}
+
+			// try to use user provided format
+			if (customFormatValue != null) {
+				try {
+					format = new DecimalFormat(customFormatValue);
+				} catch (IllegalArgumentException e) {
+					Skript.error("Invalid number format: " + customFormatValue);
+					return false;
+				}
+			}
+		// if not set then assign default format
+		} else if (customFormat == null) {
+			format = DEFAULT_FORMAT;
+		}
+		
+		return true;
+	}
+
+	@Override
+	protected String[] get(Event event, Number[] source) {
+		DecimalFormat format;
+		String formatString;
+
+		if (customFormat != null && this.format == null) { // customFormat is not Literal or VariableString
+			formatString = customFormat.getSingle(event);
+			if (formatString == null)
+				return new String[0];
+
+			try {
+				format = new DecimalFormat(formatString);
+			} catch (IllegalArgumentException e) {
```

Comment:
Should we give a user some kind of warning if their decimal format is illegal, rather than just giving nothing?

Reply:
Sadly this will spam their console which I remember when we had a team discussion we settled on this

Reply:
Solved after converting to a function.